### PR TITLE
[waiting for author] Map class type parameters to superclass when checking class method use

### LINKED ISSUE #1337: `@classmethod` of `Generic` class
``` python
from typing import TypeVar, Generic

T = TypeVar('T')


class A(Generic[T]):
    @classmethod
    def func(cls) -> T:
        pass


class B(A['B']):
    pass


def except_b(b: B):
    pass


except_b(B.func())
```

is a valid python3 code, but mypy warns that

```
file:20: error: Argument 1 to "except_b" has incompatible type "T"; expected "B"
```

but the resolved pseudo-code of class B should be

``` python
class B(A['B']):
    @classmethod
    def func(cls) -> 'B':
        pass
```

if I replace it, mypy doesn't complain anymore.

I'm might have overlook some specificities of `@classmethod` but I think that `func` is also bind to `class B` thus should be type checking.

### PR DESCRIPTION
**Pull request by rwbarton**
Fixes #1337.

### CODE REVIEW

File: test-data/unit/check-generics.test [hunk 1]
```diff
@@ -188,6 +188,71 @@ main:9: error: Incompatible types in assignment (expression has type A[C, B], va
...
+from typing import TypeVar, Generic
+T = TypeVar('T')
+U = TypeVar('U')
+class X(Generic[T,U]):
+    def __init__(self, t: T, u: U) -> None:
+        pass
+
+class A(Generic[T]):
+    @classmethod
+    def f(cls, x: T) -> T:
```
Comment (JukkaL) [inline test-data/unit/check-generics.test]: What's the type of `cls` here?

Comment (JukkaL) [inline test-data/unit/check-generics.test]: Ah, the above question probably isn't relevant here.

Comment (rwbarton) [inline test-data/unit/check-generics.test]: It's apparently the full type of `A`, namely ``def [T] () -> sm3.A[T`1]``. That `T` is not the same as the `T` in this method, apparently: `a = cls() # type: A[int]` is accepted here. This is maybe a little weird, but not changed by this commit...

**Comment by JukkaL** [comment]
Hmm I'd like to see a more complete (approximation of) real-world example where this is being used, and I'd like it to type check. Here is a naive try (haven't tried this yet):

```
from typing import TypeVar, Generic

T = TypeVar('T')

class Maker(Generic[T]):
    @classmethod
    def make(cls) -> T:   
        # this has to be abstract, as T varies totally freely here. but i think that
        # abstract class methods aren't supported. this is close enough.
        raise NotImplementedError

class B(Maker['B']):
    @classmethod
    def make(cls) -> B:
        return B()  # ok

class C(Maker['C']):
    @classmethod
    def make(cls) -> int:   # should be an error
        return 1

def f(o: Maker[T]) -> T:
    return o.make()  # should be okay?

b = f(B())  # type: B  # okay

def g(t: Type[Maker[T]]) -> T:   # should this work?
    return t.make()

bb = g(B)  # type: B  # okay
```

**Comment by gvanrossum** [comment]
@rwbarton Do you have time to address Jukka's suggestion above?

**Comment by gvanrossum** [comment]
Ping

**Comment by gvanrossum** [comment]
Hey Reid,

It looks like this PR is pretty much stuck. I'm trying to clean up stuck PRs so that the list of PRs is not too long. If you feel like working on this some more, just go ahead and reopen it, or open a new PR -- the corresponding issue will remain open with a reference to this (closed) PR in it, so if someone else feels like tackling this problem they still have this PR as a starting point.

**Pull request closed without merging**